doc/dbdctl.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DBDCTL" "8" "March 2016" "Datto Inc" ""
.
.SH "NAME"
\fBdbdctl\fR
.
.SH "NAME"
dattobd \- Control the Datto block device kernel module\.
.
.SH "SYNOPSIS"
\fBdbdctl <sub\-command> [<args>]\fR
.
.SH "DESCRIPTION"
\fBdbdctl\fR is the userspace tool used to manage the dattobd kernel module\. It provides an interface to create, delete, reload, transition, and configure on\-disk snapshots and certain parameters of the kernel module itself\.
.
.P
This manual page describes \fBdbdctl\fR briefly\. More detail is available in the Git repository located at https://github\.com/datto/dattobd\.
.
.SH "OPTIONS"
.
.nf

\-c cache\-size
     Specify how big the in\-memory data cache can grow to (in MB)\. Defaults to 300 MB\.

\-f fallocate
     Specify the maximum size of the COW file on disk\.
.
.fi
.
.SH "SUB\-COMMANDS"
.
.SS "setup\-snapshot"
\fBdbdctl setup\-snapshot [\-c <cache size>] [\-f <fallocate>] <block device> <cow file path> <minor>\fR
.
.P
Sets up a snapshot of \fB<block device>\fR, saving all COW data to \fB<cow file path>\fR\. The snapshot device will be \fB/dev/datto<minor>\fR\. The minor number will be used as a reference number for all other \fBdbdctl\fR commands\. \fB<cow file path>\fR must be a path on the \fB<block device>\fR\.
.
.SS "reload\-snapshot"
\fBdbdctl reload\-snapshot [\-c <cache size>] <block device> <cow file> <minor>\fR
.
.P
Reloads a snapshot\. This command is meant to be run before the block device is mounted, after a reboot or after the driver is unloaded\. It notifies the kernel driver to expect the block device specified to come back online\. This command requires that the snapshot was cleanly unmounted in snapshot mode beforehand\. If this is not the case, the snapshot will be put into the failure state once it attempts to come online\. The minor number will be used as a reference number for all other \fBdbdctl\fR commands\.
.
.SS "reload\-incremental"
\fBdbdctl reload\-incremental [\-c <cache size>] <block device> <cow file> <minor>\fR
.
.P
Reloads a block device that was in incremental mode\. See \fBreload\-snapshot\fR for restrictions\.
.
.SS "transition\-to\-incremental"
\fBdbdctl transition\-to\-incremental <minor>\fR
.
.P
Transitions a snapshot COW file to incremental mode, which only tracks which blocks have changed since a snapshot started\. This will remove the associated snapshot device\.
.
.SS "transition\-to\-snapshot"
\fBdbdctl transition\-to\-snapshot [\-f <fallocate>] <cow file> <minor>\fR
.
.P
Transitions a block device in incremental mode to snapshot mode\. This call ensures no writes are missed between tearing down the incremental and setting up the new snapshot\. The new snapshot data will be recorded in \fB<cow file>\fR\. The old cow file will still exist after this and can be used to efficiently copy only changed blocks using a tool succh as \fBupdate\-img\fR\.
.
.SS "destroy"
\fBdbdctl destroy <minor>\fR
.
.P
Cleanly and completely removes the snapshot or incremental, unlinking the associated COW file\.
.
.SS "reconfigure"
\fBdbdctl reconfigure [\-c <cache size>] <minor>\fR
.
.P
Allows you to reconfigure various parameters of a snapshot while it is online\. Currently only the index cache size (given in MB) can be changed dynamically\.
.
.SS "expand-cow-file"
\fBdbdctl expand-cow-file <size> <minor>\fR
.
.P
Expands cow file in snapshot mode by size (given in megabytes)\.
.
.SS "reconfigure-auto-expand"
\fBdbdctl reconfigure-auto-expand [-r <reserved space>] <step size> <minor>\fR
.
.P
Enable auto-expand of cow file in snapshot mode by <step size> (given in megabytes). Auto-expand works in that way that at least <reserved space> (given in megabytes) is left available after each step for regular users of filesystem\.
.
.SS "EXAMPLES"
\fB# dbdctl setup\-snapshot /dev/sda1 /var/backup/datto 4\fR
.
.P
This command will set up a new COW snapshot device tracking \fB/dev/sda1\fR at \fB/dev/datto4\fR\. This block device is backed by a new file created at the path \fB/var/backup/datto\fR\.
.
.P
\fB# dbdctl transition\-to\-incremental 4\fR
.
.P
Transitions the snapshot specified by the minor number to incremental mode\.
.
.P
\fB# dbdctl transition\-to\-snapshot /var/backup/datto1 4\fR
.
.P
Cleanly transitions the incremental to a new snapshot, using \fB/var/backup/datto1\fR as the new COW file\. At this point a second backup can be taken, either doing a full copy with a tool like \fBdd\fR or an incremental copy using a tool such as \fBupdate\-img\fR, if a previous snapshot backup exists\.
.
.P
\fB# dbdctl reconfigure \-c 400 4\fR
.
.P
Reconfigures the block device to have an in\-memory index cache size of 400 MB\.
.
.P
\fB# dbdctl destroy 4\fR
.
.P
This will stop tracking \fB/dev/sda1\fR, remove the associated \fB/dev/datto4\fR (since the device is in snapshot mode), delete the COW file backing it, and perform all other cleanup\.
.
.P
\fB# dbdctl reload\-snapshot /dev/sda1 /var/backup/datto1 4\fR
.
.P
After a reboot, this command may be performed in the early stages of boot, before the block device is mounted read\-write\. This will notify the driver to expect a block device \fB/dev/sda1\fR that was left in snapshot mode to come online with a COW file located at \fB/var/backup/datto1\fR (relative to the mountpoint), and that the reloaded snapshot should come online at minor number 4\. If a problem is discovered when the block device comes online, this block device will be put into the failure state, which will be reported in \fB/proc/datto\-info\fR
.
.P
\fB# dbdctl reload\-incremental /dev/sda5 /var/backup/datto1 4\fR
.
.P
This will act the same as \fBreload\-snapshot\fR, but for a device that was left in incremental mode\.
.
.SH "Bugs"
.
.SH "Author"
.
.nf

Tom Caputi (tcaputi@datto\.com)
.
.fi



app/dbdctl.c
// SPDX-License-Identifier: GPL-2.0-only

/*
 * Copyright (C) 2015 Datto Inc.
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include <ctype.h>

#include "libdattobd.h"

static void print_help(int status){
	printf("Usage:\n");
	printf("\tdbdctl setup-snapshot [-c <cache size>] [-f fallocate] <block device> <cow file> <minor>\n");
	printf("\tdbdctl reload-snapshot [-c <cache size>] <block device> <cow file> <minor>\n");
	printf("\tdbdctl reload-incremental [-c <cache size>] <block device> <cow file> <minor>\n");
	printf("\tdbdctl destroy <minor>\n");
	printf("\tdbdctl transition-to-incremental <minor>\n");
	printf("\tdbdctl transition-to-snapshot [-f fallocate] <cow file> <minor>\n");
	printf("\tdbdctl reconfigure [-c <cache size>] <minor>\n");
	printf("\tdbdctl expand-cow-file <size> <minor>\n");
	printf("\tdbdctl reconfigure-auto-expand [-r <reserved space>] <step size> <minor>\n");
	printf("\tdbdctl help\n\n");
	printf("<cow file> should be specified as an absolute path.\n");
	printf("cache size should be provided in bytes, and fallocate should be provided in megabytes.\n");
	printf("in expand-cow-file and reconfigure-auto-expand size should be provided in megabytes.\n");
	printf("note: if the -c or -f options are not specified for any given call, module defaults are used.\n");
	exit(status);
}

static int parse_ul(const char *str, unsigned long *out){
	long long tmp;
	const char *c = str;

	//check that string is an integer number and has a length
	do{
		if(!isdigit(*c)) goto error;
		c++;
	}while(*c);

	//convert to long long
	tmp = strtoll(str, NULL, 0);
	if(errno) goto error;

	//check boundaries
	if(tmp < 0 || tmp == LLONG_MAX){
		errno = ERANGE;
		goto error;
	}

	*out = (unsigned long)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int parse_l(const char *str, long *out){
	long long tmp;
	int first;
	const char *c = str;

	//check that string is an integer number and has a length
	first = 1;
	do{
		if(!isdigit(*c) && (!first || *c != '-')) goto error;
		first = 0;
		c++;
	}while(*c);

	//convert to long long
	tmp = strtoll(str, NULL, 0);
	if(errno) goto error;

	//check boundaries
	if(tmp < (long long)LONG_MIN || tmp > (long long)LONG_MAX){
		errno = ERANGE;
		goto error;
	}

	*out = (long)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int parse_ui(const char *str, unsigned int *out){
	long tmp;
	const char *c = str;

	//check that string is an integer number and has a length
	do{
		if(!isdigit(*c)) goto error;
		c++;
	}while(*c);

	//convert to long
	tmp = strtol(str, NULL, 0);
	if(errno) goto error;

	//check boundaries
	if(tmp < 0 || tmp == LONG_MAX){
		errno = ERANGE;
		goto error;
	}

	*out = (unsigned int)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int parse_ui64(const char *str, uint64_t *out){
	if(sizeof(unsigned long long) < sizeof(uint64_t)){
		errno = EINVAL;
		goto error;
	}
	
	unsigned long long tmp;
	const char *c = str;

	//check that string is an integer number and has a length
	do{
		if(!isdigit(*c)) goto error;
		c++;
	}while(*c);

	//convert to unsigned long long
	tmp = strtoull(str, NULL, 0);
	if(errno) goto error;

	*out = (uint64_t)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int handle_setup_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0, fallocated_space = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:f:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		case 'f':
			ret = parse_ul(optarg, &fallocated_space);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return dattobd_setup_snapshot(minor, bdev, cow, fallocated_space, cache_size);

error:
	perror("error interpreting setup snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reload_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return dattobd_reload_snapshot(minor, bdev, cow, cache_size);

error:
	perror("error interpreting reload snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reload_inc(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return dattobd_reload_incremental(minor, bdev, cow, cache_size);

error:
	perror("error interpreting reload incremental parameters");
	print_help(-1);
	return 0;
}

static int handle_destroy(int argc, char **argv){
	int ret;
	unsigned int minor;

	if(argc != 2){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[1], &minor);
	if(ret) goto error;

	return dattobd_destroy(minor);

error:
	perror("error interpreting destroy parameters");
	print_help(-1);
	return 0;
}

static int handle_transition_inc(int argc, char **argv){
	int ret;
	unsigned int minor;

	if(argc != 2){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[1], &minor);
	if(ret) goto error;

	return dattobd_transition_incremental(minor);

error:
	perror("error interpreting transition to incremental parameters");
	print_help(-1);
	return 0;
}

static int handle_transition_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long fallocated_space = 0;
	char *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "f:")) != -1){
		switch(c){
		case 'f':
			ret = parse_ul(optarg, &fallocated_space);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 2){
		errno = EINVAL;
		goto error;
	}

	cow = argv[optind];

	ret = parse_ui(argv[optind + 1], &minor);
	if(ret) goto error;

	return dattobd_transition_snapshot(minor, cow, fallocated_space);

error:
	perror("error interpreting transition to snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reconfigure(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 1){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[optind], &minor);
	if(ret) goto error;

	return dattobd_reconfigure(minor, cache_size);

error:
	perror("error interpreting reconfigure parameters");
	print_help(-1);
	return 0;
}

static int handle_expand_cow_file(int argc, char **argv){
	int ret;
	unsigned int minor;
	uint64_t size;

	if(argc != 3){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui64(argv[1], &size);
	if(ret) goto error;

	ret = parse_ui(argv[2], &minor);
	if(ret) goto error;

	return dattobd_expand_cow_file(minor, size);

error:
	perror("error interpreting expand parameters");
	print_help(-1);
	return 0;
}

static int handle_reconfigure_auto_expand(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	uint64_t step_size;
	uint64_t reserved_space = 0;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "r:")) != -1){
		switch(c){
		case 'r':
			ret = parse_ui64(optarg, &reserved_space);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 2){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui64(argv[optind], &step_size);
	ret = parse_ui(argv[optind+1], &minor);
	if(ret) goto error;

	return dattobd_reconfigure_auto_expand(minor, step_size, reserved_space);

error:
	perror("error interpreting reconfigure auto expand parameters");
	print_help(-1);
	return 0;
}

int main(int argc, char **argv){
	int ret = 0;

	//check argc
	if(argc < 2) print_help(-1);

	if(access("/dev/datto-ctl", F_OK) != 0){
		errno = EINVAL;
		perror("driver does not appear to be loaded");
		return -1;
	}

	//route to appropriate handler or print help
	if(!strcmp(argv[1], "setup-snapshot")) ret = handle_setup_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reload-snapshot")) ret = handle_reload_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reload-incremental")) ret = handle_reload_inc(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "destroy")) ret = handle_destroy(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "transition-to-incremental")) ret = handle_transition_inc(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "transition-to-snapshot")) ret = handle_transition_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reconfigure")) ret = handle_reconfigure(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "expand-cow-file")) ret = handle_expand_cow_file(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reconfigure-auto-expand")) ret = handle_reconfigure_auto_expand(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "help")) print_help(0);
	else print_help(-1);

	if(ret) perror("driver returned an error performing specified action. check dmesg for more info");

	return ret;
}
